### DISCUSSION TITLE
Won't boot to desktop after Debian 12.6 update

### DISCUSSION THREAD

**Discussion by CotonDeCoolear** [SpiralLinux/SpiralLinux-project#448] [Q&A] [answered]
I've been using Debian 12 (KDE version) via Spiral Linux for a week or so now, and I've been thrilled with it. Thanks so much for creating this sensibly-configured Debian distro!

I've suddenly run into a big problem, though. I installed the just-released Debian 12.6 update (or at least I think I did, there were around 100 packages pending in Discover), restarted, and... just got a white underscore blinking on a black screen.

I restarted again and got the same result, though eventually this message came up. Unfortunatelty, it just stayed there. The boot process never continued.
```
Finished plymouth-quit.service - Terminate Plymouth Boot screen.
Finished plymouth-quit-wait.service - Hold until boot process finishes up.
```

Fortunately, everything is fine after using Snapper to rollback to a pre-update snapshot. So it's definitely the update that broke something critical with my SpiralLinux install.

I think I'm running a pretty "stock" installation of SpiralLinux Debian 12 stable KDE. I'm using kernel 6.7.12+bpo-amd64, which I believe the SpiralLinux installer picked out for me.

Any ideas what could be causing the Debian 12.6 update to prevent me from booting? Thanks so much!

**Comment by geckolinux** [comment]
Hi there, could you please try booting with the previously installed working kernel version from Debian Backports, or the 6.1.0 kernel from Debian Stable that is also pre-installed in SpiralLinux?

**Reply by CotonDeCoolear** [reply]
Thanks so much for getting back to me! Trying the 6.1.0 Debian Stable kernel was my first instinct too, but it doesn't seem to appear as an option at startup in GRUB. The only kernels that appear there are:

Debian (SpiralLinux Plasma) GNU/Linux, with Linux 6.7.12+bpo-amd64
Debian (SpiralLinux Plasma) GNU/Linux, with Linux 6.7.12+bpo-amd64 (recovery mode)
Debian (SpiralLinux Plasma) GNU/Linux, with Linux 6.5.0-0.deb12.1-amd64
Debian (SpiralLinux Plasma) GNU/Linux, with Linux 6.5.0-0.deb12.1-amd64 (recovery mode)

However, if I look in /lib/modules then 6.1.0-13-amd64 is right there with the other two listed above.

Any idea how I can get my system to see/boot from the 6.1 stable kernel? Given my goals for this system (stability), that's probably the one I'd prefer to be on anyway.

Again, thanks so much for your help!

**Reply by geckolinux** [reply, marked as answer]
Hi there, does it boot with kernel 6.5.0 ?

**Reply by CotonDeCoolear** [reply]
I'm very pleased to say that it does boot with kernel 6.5.0! Thank you so much for your remarkably prompt, helpful advice.

**Reply by CotonDeCoolear** [reply]
Everything works perfectly with kernel 6.5.0 now; thanks again!

If it's not too much bother, what would be involved with getting it to boot from kernel 6.1.0? Sorry if I'm being daft. Just trying to plan for a proper stable Debian future. Again, thanks so much for all of your help.

**Reply by geckolinux** [reply]
Hi there, please try this:
1. Delete `/etc/apt/preferences.d/99backports`
2. In Synaptic search for `linux-image-amd64` and set **Package > Force version...** to `6.1.94-1 (stable)` and apply changes
3. Reboot and select the 6.1.0 kernel
4. If everything works OK use Synaptic to search for and delete any installed packages that contain `linux-image-6.5` or `linux-image-6.7`

**Reply by CotonDeCoolear** [reply]
Hmm. Because the `6.5.0` kernel has been working perfectly, I tried doing this a little bit out of order. So that I could have both the 6.1 stable kernel available as an option but retain the 6.5.0 backport kernel as it's giving me no trouble.

As such, I only did steps 2 and 3.

Sadly, after rebooting and select the `6.1.0` kernel, the computer boots the same way it did with the `6.7.12` kernel: to a white underscore on a black screen. Rebooting and selecting the `6.5.0` kernel works as expected.

Any idea what could be causing this? I will note that Synaptic very briefly showed a few warnings about potentially missing amdgpu packages/firmware (I have a Radeon RX 7900 XT) when I had it "Force Version" `6.1.94-1 (stable)`.

Again, thank you so much for all of your help (and for creating such a great advanced Debian installer). And sorry in advance if I caused this trouble myself by skipping steps 1 and 4 of your helpful list; I just didn't want to end up in a situation where I had no working kernel at all. But, most of all, thanks for making this great Debian version!

**Reply by geckolinux** [reply]
Hi there, every kernel version has different improvements and regressions on different hardware. So in your case you can remove all the kernel packages except version 6.5, as well as the `linux-image-amd64` metapackage so that it doesn't try to upgrade it to the latest version.

> I will note that Synaptic very briefly showed a few warnings about potentially missing amdgpu packages/firmware

That's a standard unimportant warning.

**Comment by MartinHemma** [comment]
I may have the same problem on one of my three Spiral Linux installations. I am not sure though. My situation:

1. Main computer will not boot automagically after recent update. The default kernel is 6.9.7+bpo-amd64. Using an older kernel (6.7.12+bpo-amd64) works fine. `apt upgrade` returns error message saying something (in Swedish) about dependency issue for linux-headers-amd64.
2. Side-kick computer (Lenovo X230) is happily booting with kernel 6.9.7+bpo-amd64. `apt update`, `apt upgrade` runs without hitches.
3. Virtual machine (Qemu) boots with kernel 6.1.0-23-amd64.

All three systems have the exact same `/etc/apt/sources.list`. Running `apt update` on all systems results in messages telling me all packages are up to date.

How do I repair my main system?

Why is the virtual machine using a different kernel? I have not tried to influence the kernel choice in any of these systems.

TIA

/Martin

**Reply by marwenius** [reply]
The problems may be resolved by reinstalling.

I would try the following command:
`sudo apt --reinstall install -t bookworm-backports linux-image-amd64 linux-headers-amd64`

If this does not work, I would try this:
`sudo apt -f install`

Are any Nvidia drivers installed? These may need to be uninstalled first before the kernel update is carried out. They can then be reinstalled.

**Reply by geckolinux** [reply]
> Why is the virtual machine using a different kernel? I have not tried to influence the kernel choice in any of these systems.

I assume this is when booting the live system? In that case it's due to a different live system bootloader being used when the system is running in BIOS mode vs. UEFI.

**Reply by MartinHemma** [reply]
@marwenius Neither command made any difference. :-(

@geckolinux Sorry, I neglected to point out that the virtual system does not bother to install the kernels the real systems use (try to use). Hence, my mentioning of the sources.list files being identical. The virtual system is older than the real systems. I think it was installed using an older ISO.

/Martin

**Reply by marwenius** [reply]
@MartinHemma:

Are any error messages displayed when executing the commands? If so, could you post them here? That would make troubleshooting easier.

Otherwise you could also [roll back to an earlier snapshot](https://github.com/SpiralLinux/SpiralLinux-project/wiki#working-with-btrfs-snapshots-and-rollbacks) and start again from there?

**Reply by MartinHemma** [reply]
```
(base) martin@apollo:~$ sudo apt --reinstall install -t bookworm-backports linux-image-amd64 linux-headers-amd64
[sudo] lösenord för martin:         
Läser paketlistor… Färdig
Bygger beroendeträd… Färdig
Läser tillståndsinformation… Färdig
The following additional packages will be installed:
  hyperv-daemons libcpupower1 linux-cpupower linux-image-6.9.7+bpo-amd64 linux-libc-dev
Föreslagna paket:
  linux-doc-6.9 debian-kernel-handbook grub-pc | grub-efi-amd64 | extlinux
Rekommenderade paket:
  apparmor
Följande paket kommer att TAS BORT:
  linux-image-6.9.7+bpo-amd64-unsigned
Följande NYA paket kommer att installeras:
  linux-image-6.9.7+bpo-amd64 linux-image-amd64
Följande paket kommer att uppgraderas:
  hyperv-daemons libcpupower1 linux-cpupower linux-libc-dev
4 att uppgradera, 2 att nyinstallera, 1 att installera om, 1 att ta bort och 136 att inte uppgradera.
3 är inte helt installerade eller borttagna.
Behöver hämta 0 B/105 MB arkiv.
Efter denna åtgärd kommer ytterligare 3 505 kB utrymme användas på disken.
Vill du fortsätta? [J/n] y
E: Internal Error, No file name for linux-headers-amd64:amd64`
```

And

```
(base) martin@apollo:~$ sudo apt -f install
Läser paketlistor… Färdig
Bygger beroendeträd… Färdig
Läser tillståndsinformation… Färdig
0 att uppgradera, 0 att nyinstallera, 0 att ta bort och 0 att inte uppgradera.
3 är inte helt installerade eller borttagna.
Efter denna åtgärd kommer ytterligare 0 B utrymme användas på disken.
Ställer in linux-image-6.9.7+bpo-amd64-unsigned (6.9.7-1~bpo12+1) ...
/etc/kernel/postinst.d/dkms:
dkms: running auto installation service for kernel 6.9.7+bpo-amd64.
Sign command: /lib/modules/6.9.7+bpo-amd64/build/scripts/sign-file
Signing key: /var/lib/dkms/mok.key
Public certificate (MOK): /var/lib/dkms/mok.pub

Building module:
Cleaning build area...
make -j20 KERNELRELEASE=6.9.7+bpo-amd64 -C /lib/modules/6.9.7+bpo-amd64/build M=/var/lib/dkms/virtualbox/7.0.12/build...(bad exit status: 2)
Error! Bad return status for module build on kernel: 6.9.7+bpo-amd64 (x86_64)
Consult /var/lib/dkms/virtualbox/7.0.12/build/make.log for more information.
Error! One or more modules failed to install during autoinstall.
Refer to previous errors for more information.
dkms: autoinstall for kernel: 6.9.7+bpo-amd64 failed!
run-parts: /etc/kernel/postinst.d/dkms exited with return code 11
dpkg: fel vid hantering av paketet linux-image-6.9.7+bpo-amd64-unsigned (--configure):
 underprocessen installerat paket linux-image-6.9.7+bpo-amd64-unsigned:s skript post-installation gav felkod 1 vid avslut
dpkg: beroendeproblem förhindrar konfigurering av linux-headers-6.9.7+bpo-amd64:
 linux-headers-6.9.7+bpo-amd64 är beroende av linux-image-6.9.7+bpo-amd64 (= 6.9.7-1~bpo12+1) | linux-image-6.9.7+bpo-amd64-unsigned (= 6.9.7-1
~bpo12+1), men:
  Paketet linux-image-6.9.7+bpo-amd64 är ej installerat.
  Paketet linux-image-6.9.7+bpo-amd64-unsigned har inte konfigurerats ännu.

dpkg: fel vid hantering av paketet linux-headers-6.9.7+bpo-amd64 (--configure):
 beroendeproblem - lämnar okonfigurerad
dpkg: beroendeproblem förhindrar konfigurering av linux-headers-amd64:
 linux-headers-amd64 är beroende av linux-headers-6.9.7+bpo-amd64 (= 6.9.7-1~bpo12+1), men:
  Paketet linux-headers-6.9.7+bpo-amd64 har inte konfigurerats ännu.

dpkg: fel vid hantering av paketet linux-headers-amd64 (--configure):
 beroendeproblem - lämnar okonfigurerad
Fel uppstod vid hantering:
 linux-image-6.9.7+bpo-amd64-unsigned
 linux-headers-6.9.7+bpo-amd64
 linux-headers-amd64
Config används.
E: Sub-process /usr/bin/dpkg returned an error code (1)`
```
Partially (!) in Swedish but maybe not that hard to decipher.

Yes, roll back should work I guess. I have never tried that before. It will not tell me what went wrong. That will be for later.

/Martin

**Reply by MartinHemma** [reply]
Hmm, copy/paste and typography did not work as expected :-(
/Martin

**Reply by marwenius** [reply]
Thanks for the console output.

I would try to remove the 6.9 kernel and reinstall it:
```
sudo apt remove linux-image-6.9.* linux-headers-6.9.*
sudo apt autoremove
sudo apt install -t bookworm-backports linux-image-amd64 linux-headers-amd64
```
And since the one error also has something to do with Virtualbox, I would also reinstall this package. Perhaps even with the latest version from the official homepage:
```
sudo apt remove virtualbox
sudo apt autoremove
wget -c https://download.virtualbox.org/virtualbox/7.0.20/virtualbox-7.0_7.0.20-163906~Debian~bookworm_amd64.deb
wget -c https://download.virtualbox.org/virtualbox/7.0.20/Oracle_VM_VirtualBox_Extension_Pack-7.0.20.vbox-extpack
sudo dpkg -i virtualbox-7.0_7.0.20-163906~Debian~bookworm_amd64.deb
sudo rcvboxdrv setup
```

**Reply by MartinHemma** [reply]
I have just tried the first three commands. The first two do not generate error messages as far as I can see. The third, however, still complains and now I also see Virtualbox seems to be involved.

```
(base) martin@apollo:~$ sudo apt install -t bookworm-backports linux-image-amd64 linux-headers-amd64
Läser paketlistor… Färdig
Bygger beroendeträd… Färdig
Läser tillståndsinformation… Färdig
The following additional packages will be installed:
  hyperv-daemons libcpupower1 linux-cpupower linux-headers-6.9.7+bpo-amd64 linux-headers-6.9.7+bpo-common
  linux-image-6.9.7+bpo-amd64 linux-kbuild-6.9.7+bpo linux-libc-dev
Föreslagna paket:
  linux-doc-6.9 debian-kernel-handbook grub-pc | grub-efi-amd64 | extlinux
Rekommenderade paket:
  apparmor
Följande NYA paket kommer att installeras:
  linux-headers-6.9.7+bpo-amd64 linux-headers-6.9.7+bpo-common linux-headers-amd64 linux-image-6.9.7+bpo-amd64
  linux-image-amd64 linux-kbuild-6.9.7+bpo
Följande paket kommer att uppgraderas:
  hyperv-daemons libcpupower1 linux-cpupower linux-libc-dev
4 att uppgradera, 6 att nyinstallera, 0 att ta bort och 136 att inte uppgradera.
Behöver hämta 13,4 MB/119 MB arkiv.
Efter denna åtgärd kommer ytterligare 176 MB utrymme användas på disken.
Vill du fortsätta? [J/n] y
Läs:1 https://deb.debian.org/debian bookworm-backports/main amd64 linux-headers-6.9.7+bpo-common all 6.9.7-1~bpo12+1 [10,5 MB]
Läs:2 https://deb.debian.org/debian bookworm-backports/main amd64 linux-kbuild-6.9.7+bpo amd64 6.9.7-1~bpo12+1 [1 094 kB]
Läs:3 https://deb.debian.org/debian bookworm-backports/main amd64 linux-headers-6.9.7+bpo-amd64 amd64 6.9.7-1~bpo12+1 [1 837 kB]
Läs:4 https://deb.debian.org/debian bookworm-backports/main amd64 linux-headers-amd64 amd64 6.9.7-1~bpo12+1 [1 416 B]
Hämtade 13,4 MB på 0s (29,9 MB/s)                   
Läser förändringsloggar... Klar
(Läser databasen ... 395968 filer och kataloger installerade.)
Förbereder att packa upp .../0-hyperv-daemons_6.9.7-1~bpo12+1_amd64.deb ...
Packar upp hyperv-daemons (6.9.7-1~bpo12+1) över (6.1.99-1) ...
Förbereder att packa upp .../1-libcpupower1_6.9.7-1~bpo12+1_amd64.deb ...
Packar upp libcpupower1 (6.9.7-1~bpo12+1) över (6.1.99-1) ...
Förbereder att packa upp .../2-linux-cpupower_6.9.7-1~bpo12+1_amd64.deb ...
Packar upp linux-cpupower (6.9.7-1~bpo12+1) över (6.1.99-1) ...
Väljer tidigare ej valt paket linux-headers-6.9.7+bpo-common.
Förbereder att packa upp .../3-linux-headers-6.9.7+bpo-common_6.9.7-1~bpo12+1_all.deb ...
Packar upp linux-headers-6.9.7+bpo-common (6.9.7-1~bpo12+1) ...
Väljer tidigare ej valt paket linux-image-6.9.7+bpo-amd64.
Förbereder att packa upp .../4-linux-image-6.9.7+bpo-amd64_6.9.7-1~bpo12+1_amd64.deb ...
Packar upp linux-image-6.9.7+bpo-amd64 (6.9.7-1~bpo12+1) ...
Väljer tidigare ej valt paket linux-kbuild-6.9.7+bpo.
Förbereder att packa upp .../5-linux-kbuild-6.9.7+bpo_6.9.7-1~bpo12+1_amd64.deb ...
Packar upp linux-kbuild-6.9.7+bpo (6.9.7-1~bpo12+1) ...
Väljer tidigare ej valt paket linux-headers-6.9.7+bpo-amd64.
Förbereder att packa upp .../6-linux-headers-6.9.7+bpo-amd64_6.9.7-1~bpo12+1_amd64.deb ...
Packar upp linux-headers-6.9.7+bpo-amd64 (6.9.7-1~bpo12+1) ...
Väljer tidigare ej valt paket linux-headers-amd64.
Förbereder att packa upp .../7-linux-headers-amd64_6.9.7-1~bpo12+1_amd64.deb ...
Packar upp linux-headers-amd64 (6.9.7-1~bpo12+1) ...
Väljer tidigare ej valt paket linux-image-amd64.
Förbereder att packa upp .../8-linux-image-amd64_6.9.7-1~bpo12+1_amd64.deb ...
Packar upp linux-image-amd64 (6.9.7-1~bpo12+1) ...
Förbereder att packa upp .../9-linux-libc-dev_6.9.7-1~bpo12+1_all.deb ...
Packar upp linux-libc-dev (6.9.7-1~bpo12+1) över (6.1.99-1) ...
Ställer in linux-libc-dev (6.9.7-1~bpo12+1) ...
Ställer in linux-image-6.9.7+bpo-amd64 (6.9.7-1~bpo12+1) ...
I: /vmlinuz.old is now a symlink to boot/vmlinuz-6.1.0-21-amd64
I: /initrd.img.old is now a symlink to boot/initrd.img-6.1.0-21-amd64
I: /vmlinuz is now a symlink to boot/vmlinuz-6.9.7+bpo-amd64
I: /initrd.img is now a symlink to boot/initrd.img-6.9.7+bpo-amd64
/etc/kernel/postinst.d/dkms:
dkms: running auto installation service for kernel 6.9.7+bpo-amd64.
Sign command: /lib/modules/6.9.7+bpo-amd64/build/scripts/sign-file
Signing key: /var/lib/dkms/mok.key
Public certificate (MOK): /var/lib/dkms/mok.pub

Building module:
Cleaning build area...
make -j20 KERNELRELEASE=6.9.7+bpo-amd64 KVER=6.9.7+bpo-amd64....
Signing module /var/lib/dkms/broadcom-sta/6.30.223.271/build/wl.ko
Cleaning build area...

wl.ko.xz:
Running module version sanity check.
 - Original module
   - No original module exists within this kernel
 - Installation
   - Installing to /lib/modules/6.9.7+bpo-amd64/updates/dkms/
depmod....
Sign command: /lib/modules/6.9.7+bpo-amd64/build/scripts/sign-file
Signing key: /var/lib/dkms/mok.key
Public certificate (MOK): /var/lib/dkms/mok.pub

Building module:
Cleaning build area...
make -j20 KERNELRELEASE=6.9.7+bpo-amd64 -C /lib/modules/6.9.7+bpo-amd64/build M=/var/lib/dkms/virtualbox/7.0.12/buil
d...(bad exit status: 2)
Error! Bad return status for module build on kernel: 6.9.7+bpo-amd64 (x86_64)
Consult /var/lib/dkms/virtualbox/7.0.12/build/make.log for more information.
Error! One or more modules failed to install during autoinstall.
Refer to previous errors for more information.
dkms: autoinstall for kernel: 6.9.7+bpo-amd64 failed!
run-parts: /etc/kernel/postinst.d/dkms exited with return code 11
dpkg: fel vid hantering av paketet linux-image-6.9.7+bpo-amd64 (--configure):
 underprocessen installerat paket linux-image-6.9.7+bpo-amd64:s skript post-installation gav felkod 1 vid avslut
Ställer in linux-headers-6.9.7+bpo-common (6.9.7-1~bpo12+1) ...
Ställer in libcpupower1 (6.9.7-1~bpo12+1) ...
Ställer in linux-cpupower (6.9.7-1~bpo12+1) ...
dpkg: beroendeproblem förhindrar konfigurering av linux-headers-6.9.7+bpo-amd64:
 linux-headers-6.9.7+bpo-amd64 är beroende av linux-image-6.9.7+bpo-amd64 (= 6.9.7-1~bpo12+1) | linux-image-6.9.7+bp
o-amd64-unsigned (= 6.9.7-1~bpo12+1), men:
  Paketet linux-image-6.9.7+bpo-amd64 har inte konfigurerats ännu.
  Paketet linux-image-6.9.7+bpo-amd64-unsigned är ej installerat.

dpkg: fel vid hantering av paketet linux-headers-6.9.7+bpo-amd64 (--configure):
 beroendeproblem - lämnar okonfigurerad
Ställer in hyperv-daemons (6.9.7-1~bpo12+1) ...
Ställer in linux-kbuild-6.9.7+bpo (6.9.7-1~bpo12+1) ...
dpkg: beroendeproblem förhindrar konfigurering av linux-headers-amd64:
 linux-headers-amd64 är beroende av linux-headers-6.9.7+bpo-amd64 (= 6.9.7-1~bpo12+1), men:
  Paketet linux-headers-6.9.7+bpo-amd64 har inte konfigurerats ännu.

dpkg: fel vid hantering av paketet linux-headers-amd64 (--configure):
 beroendeproblem - lämnar okonfigurerad
dpkg: beroendeproblem förhindrar konfigurering av linux-image-amd64:
 linux-image-amd64 är beroende av linux-image-6.9.7+bpo-amd64 (= 6.9.7-1~bpo12+1), men:
  Paketet linux-image-6.9.7+bpo-amd64 har inte konfigurerats ännu.

dpkg: fel vid hantering av paketet linux-image-amd64 (--configure):
 beroendeproblem - lämnar okonfigurerad
Hanterar utlösare för libc-bin (2.36-9+deb12u7) ...
Hanterar utlösare för man-db (2.11.2-2) ...
Fel uppstod vid hantering:
 linux-image-6.9.7+bpo-amd64
 linux-headers-6.9.7+bpo-amd64
 linux-headers-amd64
 linux-image-amd64
Config används.
E: Sub-process /usr/bin/dpkg returned an error code (1)
```

Virtualbox is only installed on my main machine.

Enough for tonight.

/Martin

**Reply by marwenius** [reply]
I installed SpiralLinux in Hyper-V (Windows) as a test and then executed the following commands:
```
sudo apt update
sudo apt upgrade --without-new-pkgs
sudo apt install virtualbox
sudo apt upgrade
```
The last command executes the kernel updates and it cancelled at some point with the same error pattern as yours.

I think I have found the error: It is due to the version of Virtualbox provided by Fasttrack: 7.0.12.

Kernel 6.9 is only supported from version 7.0.16 onwards:
https://www.virtualbox.org/wiki/Changelog#v16
https://www.virtualbox.org/ticket/22033

I would therefore remove the Fasttrack package as already suggested and obtain the latest version directly from Virtualbox:
```
sudo apt remove virtualbox
sudo apt autoremove
wget -c https://download.virtualbox.org/virtualbox/7.0.20/virtualbox-7.0_7.0.20-163906~Debian~bookworm_amd64.deb
sudo apt install libqt5opengl5
sudo dpkg -i virtualbox-7.0_7.0.20-163906~Debian~bookworm_amd64.deb
sudo rcvboxdrv setup
```
I had to install the libqt5opengl5 package, as Virtualbox 7.0.20 requires it.

This procedure resolved the error for me.

If you need the latest Virtualbox expansion packs:
```
wget -c https://download.virtualbox.org/virtualbox/7.0.20/Oracle_VM_VirtualBox_Extension_Pack-7.0.20.vbox-extpack
```

**Reply by MartinHemma** [reply]
Thanks a ton.

After having removed Virtualbox I re-ran the commands for removing and re-installing 6.9 kernels. Then re-booting worked like a charm.

I updated Virtualbox as per your instructions. I had to remove guest additions first.

Now, everything works smoothly again :-)

/Martin

**Comment by CotonDeCoolear** [comment]
So, I'm still getting the same basic problem (booting to a black screen with a white underscore) on every kernel other than 6.5.0, most recently 6.9.10, but I think I have some new insight into what the problem is (and how I probably caused it).

First, here's the output from `inxi -G` put into a TTY session after the 6.9.10 kernel has failed to boot:
`Device-1: AMD Navi 31 [Radeon 7900 XT/7900 XTX] driver: N/A `

When select kernel 6.5.0, the system boots properly and `inxi -G` displays:
`Device-1: AMD Navi 31 [Radeon RX 7900 XT/7900 XTX] driver: amdgpu v: kernel`

After giving the matter some thought, I think that this situation is probably my fault, but I hope someone here could be kind enough to offer me a solution if one exists. Basically, I think I "broke Debian", as they say, by very foolishly installing proprietary graphics drivers from AMD's website when I was trying to get ROCm working to replicate an AI art setup I had on my previous (Manjaro) machine. I've since noticed that the non-Debian AMD installer pinned this repo in `/etc/apt/preferences.d`:
```
Package: *
Pin: release o=repo.radeon.com
Pin-Priority: 600
```
I've since removed the files (rocm-pin-600 and repo-radeon-pin-600) responsible for creating this pin, but that doesn't appear to have changed the ability for any kernels other than 6.5.0 to boot with graphics.

So, sorry that this whole thing was probably my fault. I vow to never break Debian again! But if there's any method short of reinstallation that will recover from this error and allow kernels other than 6.5.0, I'd love to hear any suggestions. Thanks so much!

**Reply by marwenius** [reply]
Two approaches that might be helpful: 1) [get the current driver as a flatpak](https://www.reddit.com/r/linux_gaming/comments/16xpv3b/comment/k353r8x/) or 2) [build it yourself from source](https://gist.github.com/baryluk/1041204eff4cc4fad6f1508afe67b562)?

**Reply by CotonDeCoolear** [reply]
Thanks for the suggestions. I'm not sure that either of those approaches apply for my situation, however. Really, my big problem, is that Linux sees no graphics driver for my GPU at all—and thus cannot boot beyond a black screen with a white underscore—when I select any kernel other than 6.5.0. I should note that I was originally on kernel 6.7.x after first installing Spiral Linux, and that worked fine until the kernel updated with Debian 12.6. The first post in this thread details this. I believe I fell afoul of this commandment from [DontBreakDebian](https://wiki.debian.org/DontBreakDebian#Don.27t_use_GPU_manufacturer_install_scripts):

`If you absolutely must have the proprietary closed-source drivers, do not download them directly from the manufacturer's website! Installing drivers this way only works for the current kernel, and after the next kernel update, your video drivers will not work until they are manually reinstalled again.`

So, as to approach 1, that's actually what I do for gaming via Steam (and it works great), but I can only do it on kernel 6.5.0 because nothing else will get past the black screen with the white underscore. Approach 2 is intriguing, though I worry about making things even worse. Really, I'd like to know how I can get back to having Debian, with no manual hacks, appropriately install and load the GPU drivers for kernels other than 6.5.0.

I suspect that  kernel 6.5.0 was spared from disruption because it came pre-installed with the version of Spiral Linux that I'm using, so it wasn't in any way altered when I was messing around with the GPU driver install script from AMD's website.

Again, thanks for the advice. Sorry if I'm doing a terrible job of explaining this.

**Reply by marwenius** [reply]
I don't think you can avoid a bit of manual hacking.

If you know it worked well with a particular kernel, I would download and install it (and uninstall the newer kernels).

There are archives where these can be downloaded:

https://snapshot.debian.org/archive/debian/20240531T143744Z/pool/main/l/linux-signed-amd64/linux-image-6.5.0-0.deb12.1-amd64_6.5.3-1~bpo12%2B1_amd64.deb

https://snapshot.debian.org/archive/debian/20240531T143744Z/pool/main/l/linux-signed-amd64/linux-image-6.5.0-0.deb12.4-amd64_6.5.10-1~bpo12%2B1_amd64.deb

https://snapshot.debian.org/archive/debian/20240531T143744Z/pool/main/l/linux-signed-amd64/linux-image-6.7.12%2Bbpo-amd64_6.7.12-1~bpo12%2B1_amd64.deb

And then you could pin the specific version. For example:

Create the file `/etc/apt/preferences.d/kernelpin` with the following content:
```
Package: linux-image-amd64
Pin: version 6.5*
Pin priority: 1000
```

More on pinning:
- https://mike632t.wordpress.com/2023/05/24/pinning-packages-with-apt/
- https://docs.memset.com/cd/Linux-Package-Pinning.199068409.html
- https://wiki.debian.org/AptConfiguration#Other_pinning_notes


Edit:

Or take a look at the `apt-mark` command instead of pinning. Maybe you can use both together:

https://manual.siduction.org/sys-admin-apt_en.html#hold-or-downgrade-a-package

**Reply by CotonDeCoolear** [reply]
Thanks so much to everyone for all of the help you've given me with this. I'm pleased to report that I now have a much better idea of what is going on, and I've (not quite perfectly, but mostly) solved the problem.

Here, then, is the real issue:

For kernels other than 6.5.0, Debian is not loading the `amdgpu` kernel module at boot.

However, the `amdgpu` module is in fact installed and fully functional. If I select kernel `6.10.6+bpo-amd64` from GRUB, Debian will only boot to a black screen with a white underscore. However, if I switch to a TTY session and enter the command `sudo modprobe amdgpu`, the system will immediately boot to desktop and work flawlessly!

So problem basically solved! Is there a way than I can get Debian to load the `amdgpu` module on its own, without having to switch to a TTY session at boot and do it myself?

Again, thanks so much to everyone in this thread for all of the help you've offered. It feels great to be in the home stretch.

**Reply by CotonDeCoolear** [reply]
And, figured it out all the way!

There was a file called `blacklist-amdgpu.conf` hanging out in `/etc/modprobe.d`. As one would guess from the name, it was blacklisting the `amdgpu` module. After removing this file and running `sudo update-initramfs -u`, Debian now boots normally with kernel `6.10.6+bpo-amd64`.

Thanks so much to everyone for all of your help. Maybe this thread will prove useful to anyone else who winds up with the same weird, obscure problem.
